keps/sig-node/2008-forensic-container-checkpointing/README.md

```diff
@@ -140,10 +168,93 @@ continue to run without knowing a copy was created. This copy can then
+
+With the current Alpha based implementation container migration is already
+possible as documented in [Forensic Container Checkpointing Alpha][kubernetes-blog-post].
+
+The following tries to give an overview of possible use cases for
+container migration.
+
+##### Fault Tolerance
+
+Container migration for fault tolerance is one of the typical reasons to
+migrate containers or processes. It is a well researched topic especially
+in the field of high performance computing (HPC). To avoid loss of work
+already done by a container the container is migrated to another node before
+the current node crashes. There are many scientific papers describing how
+to detect a node that might soon have a hardware problem. The main goal
+of using container migration for fault tolerance is to avoid loss of already
+done work. This is, in contrast to the forensic container checkpointing use
+case, only useful for stateful containers.
+
+##### Load Balancing
+
+Container migration for load balancing is something where checkpoint/restore
+as implemented by CRIU is already used in production today. A prominent example
+is Google as presented at the Linux Plumbers conference in 2018:
+[Task Migration at Scale Using CRIU]<[task-migration]>
+
+If multiple containers are running on one node, checkpoint/restore and thus
+container migration open up the possibility to migrate containers to another
+node in case not enough resources are available on the current node. High
+priority workloads can continue to run if low priority containers are
+migrated to another node. This way, stateful low priority containers do
+not have to restart from scratch but can continue to run on another node.
+
+This might be especially interesting for AI training containers which can be
+preempted for higher priority tasks without having to restart the training
+from the start.
+
+##### Spot Instances
+
+Yet another possible use case where checkpoint/restore is already used today
+are spot instances. Spot instances are usually resources that are cheaper but
+with the drawback that they might shut down with very short notice. With the
+help of checkpoint/restore workloads on spot instances can either be
+checkpointed regularly or the checkpointing can be triggered by a signal.
+Once checkpointed the container can be moved to another instance and
+continue to run without having to start from the beginning.
+
+##### Scheduler Integration
+
+All of the above mentioned container migration use cases currently require
+manual checkpointing, manual transfer of the checkpoint archive and manual
+restore of the container (see [Forensic Container Checkpointing
+Alpha][kubernetes-blog-post] for details). If all these steps could automatically
+be performed by the scheduler this would greatly improve the user experience.
+Scheduler integration, however, is probably one use case which will be
+implemented at the very end and something for future enhancements.
```

Comment:
It might be worth noting that the scheduler could keep track of available resources in heterogenous clusters and automatically identify suitable nodes where a checkpoint can be restored (e.g., GPU nodes with MI100 vs MI210).